CRM/Core/ManagedEntities.php

```diff
@@ -523,16 +523,18 @@ private function isActivationSupported(string $entity_type): bool {
    */
   protected function getDeclarations($modules = NULL): array {
     $declarations = [];
-    // Exclude components if given a module name.
-    if (!$modules || $modules === ['civicrm']) {
-      foreach (CRM_Core_Component::getEnabledComponents() as $component) {
-        $declarations = array_merge($declarations, $component->getManagedEntities());
-      }
-    }
     CRM_Utils_Hook::managed($declarations, $modules);
     $this->validate($declarations);
-    foreach (array_keys($declarations) as $name) {
-      $declarations[$name] += ['name' => $name];
+    // FIXME: Some well-meaning developer added this a long time ago to support associative arrays
+    // that use the array index as the declaration name. But it probably never worked, because by the time it gets to this point,
+    // lots of implementations of `hook_civicrm_managed()` would have run `$declarations = array_merge($declarations, [...])`
+    // which would have reset the indexes.
```

Comment:
Wow, that's a blast from the past. I believe early iterations aspired to:

```php
function foo_civicrm_managed(&$entities) {
  $entities['my_name'] = [
    'entity' => 'Report', 
    'params' => ...,
  ];
}
```

The issue with this isn't so much changing the indexes (*like with numeric-indexes*); rather, if two modules define entities with the same name, then *one definition overwrites the other*. Simple array-keys don't provide namespace protection in this context.

You might notice that the [original and current adapters for `*.mgd.php`](https://github.com/totten/civix/commit/de4d429d4b685d970b19961c23b23ac51b4fdf0f#diff-4d4ae755aec5f28691d505f028a318ae7dcffcda7959f5b6e4a0c2fc0c6b6596R125-R130) specifically discard the array-keys. It would be really easy to keep them. Actually, in this context, array-keys could've been brought back to life while still allowing namespaces.

But then explaining the overall system would require footnotes. ("In context A, you use `name` field. In context B, you use array-key. In context C, you can use either.") It's easier to say: "Everything uses `name` field."

Reply:
> Actually, in this context, symbolic array-keys could've been brought back to life while still allowing namespaces.

@totten I think this is a classic case of `($couldWe = TRUE && $shouldWe = FALSE)`